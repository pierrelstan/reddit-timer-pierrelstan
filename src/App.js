import React from 'react';
import { BrowserRouter as Router, Switch, Route } from 'react-router-dom';
import { ThemeProvider } from 'styled-components';
import { Normalize } from 'styled-normalize';
import NoMatch from './components/NoMatch';
import Home from './pages/Home';
import Search from './pages/Search';
import GlobalStyle from './utils/GlobalStyle';
import theme from './utils/theme';

function App() {
  return (
    <ThemeProvider theme={theme}>
      <Router>
        <Normalize />
        <GlobalStyle />
        <Switch>
          <Route path="search">
            <Search />
          </Route>
          <Route exact path="/">
            <Home />
          </Route>
          <Route path="*">
            <NoMatch />
          </Route>
        </Switch>
      </Router>
    </ThemeProvider>
  );
}
export default App;
