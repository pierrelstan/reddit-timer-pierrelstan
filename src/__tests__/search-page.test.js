import React from 'react';
import {
  render, screen, waitFor, within,
} from '@testing-library/react';
import userEvent from '@testing-library/user-event';
import { BrowserRouter } from 'history';
import App from '../App';

const setup = (ui, { route = '/' } = {}) => {
  window.history.pushState({}, 'Test page', route);
  return render(ui, { wrapper: BrowserRouter });
};

describe('Testing Search Page', () => {
  test('renders error message', async () => {
    const route = '/search/failing-request';

    setup(<App />, { route });
    await waitFor(async () => {
      expect(await screen.findByText(/something went wrong/i)).toBeInTheDocument();
    });
  });

  test('loads top posts for subreddit in URL', async () => {
    const route = '/search/reactjs';

    setup(<App />, { route });

    screen.queryByText('loading-spinner.svg');

    expect(await screen.findByTestId('heatmap')).toBeInTheDocument();

    await waitFor(() => {
      expect(screen.queryByText('loading-spinner.svg')).not.toBeInTheDocument();
    });

    const heatmap = screen.getByTestId('heatmap');
    const cells = await within(heatmap).findAllByRole('button');
    expect(cells.length).toEqual(7 * 24);

    const numberOfPostsPerCell = cells.map((cell) => cell.innerHTML);
    expect(numberOfPostsPerCell).toMatchSnapshot();

    const timezone = screen.getByText('All times are shown in your timezone:');
    expect(within(timezone).getByText('Europe/Berlin')).toBeInTheDocument();
  });

  test('cell highlights on click', async () => {
    const route = '/search/reactjs';

    setup(<App />, { route });

    const heatmap = await screen.findByTestId('heatmap');
    const cells = await within(heatmap).findAllByRole('button');

    const cellToClick = cells[1];
    expect(cellToClick).toHaveStyle('border: none');

    userEvent.click(cellToClick);
    expect(cellToClick).toHaveStyle('border: 1px solid #1e2537');
  });
});

// Posts Table
async function clickFirstCellWithValue(value) {
  const heatmap = await screen.findByTestId('heatmap');
  const cell = within(heatmap).getAllByText(value)[0];
  userEvent.click(cell);
}

describe('posts table', () => {
  test('is not visible when no cell is clicked', async () => {
    const route = '/search/reactjs';

    setup(<App />, { route });
    await screen.queryByTestId('heatmap');

    expect(screen.queryByRole('table')).not.toBeInTheDocument();
  });

  test('is not visible when cell with no posts is clicked', async () => {
    const route = '/search/reactjs';

    setup(<App />, { route });
    await clickFirstCellWithValue('0');

    expect(screen.queryByRole('table')).not.toBeInTheDocument();
  });

  test('shows posts ordered by time according to cell that is clicked', async () => {
    const route = '/search/reactjs';

    setup(<App />, { route });
    await clickFirstCellWithValue('4');

    const table = screen.getByRole('table');
    const tableRows = within(table)
      .getAllByRole('row')
      .slice(1);

    const tableContent = tableRows.map((row) => {
      const cells = within(row).getAllByRole('cell');
      const titleLink = within(cells[0]).getByRole('link');
      const authorLink = within(cells[4]).getByRole('link');
      return {
        title: titleLink.innerHTML,
        href: titleLink.href,
        time: cells[1].innerHTML,
        score: cells[2].innerHTML,
        numComments: cells[3].innerHTML,
        author: authorLink.innerHTML,
        authorHref: authorLink.href,
      };
    });

    expect(tableContent).toMatchSnapshot();
  });

  test('shows no link for deleted user', async () => {
    const route = '/search/reactjs';

    setup(<App />, { route });

    const heatmap = await screen.findByTestId('heatmap');

    const sunday5pm = within(heatmap).getAllByText('7')[0];

    userEvent.click(sunday5pm);
    const table = screen.getByRole('table');

    const rowWithDeletedUser = within(table).getAllByRole('row')[2];

    const authorCell = within(rowWithDeletedUser).getAllByRole('cell')[4];

    expect(within(authorCell).queryByRole('link')).not.toBeInTheDocument();
    expect(authorCell.innerHTML).toBe('[deleted]');
  });
});
