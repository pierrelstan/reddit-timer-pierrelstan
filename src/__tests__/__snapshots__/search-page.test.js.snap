// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Testing Search Page loads top posts for subreddit in URL 1`] = `
Array [
  "2",
  "0",
  "2",
  "1",
  "2",
  "2",
  "0",
  "3",
  "0",
  "2",
  "6",
  "1",
  "4",
  "1",
  "6",
  "7",
  "6",
  "3",
  "3",
  "4",
  "4",
  "2",
  "1",
  "1",
  "3",
  "2",
  "2",
  "1",
  "1",
  "1",
  "1",
  "3",
  "5",
  "2",
  "3",
  "5",
  "5",
  "3",
  "6",
  "5",
  "6",
  "6",
  "4",
  "5",
  "4",
  "5",
  "2",
  "1",
  "1",
  "1",
  "5",
  "0",
  "1",
  "0",
  "0",
  "2",
  "1",
  "0",
  "3",
  "2",
  "2",
  "6",
  "4",
  "6",
  "10",
  "6",
  "8",
  "5",
  "1",
  "6",
  "4",
  "1",
  "1",
  "1",
  "1",
  "1",
  "2",
  "1",
  "1",
  "1",
  "3",
  "3",
  "4",
  "3",
  "2",
  "4",
  "2",
  "3",
  "4",
  "5",
  "7",
  "5",
  "4",
  "1",
  "3",
  "5",
  "3",
  "1",
  "1",
  "0",
  "0",
  "0",
  "1",
  "1",
  "3",
  "1",
  "6",
  "1",
  "3",
  "6",
  "2",
  "7",
  "2",
  "8",
  "7",
  "6",
  "1",
  "3",
  "5",
  "8",
  "2",
  "2",
  "4",
  "3",
  "1",
  "0",
  "3",
  "0",
  "1",
  "3",
  "2",
  "3",
  "1",
  "3",
  "3",
  "8",
  "7",
  "4",
  "4",
  "0",
  "7",
  "5",
  "4",
  "2",
  "0",
  "1",
  "3",
  "0",
  "4",
  "1",
  "2",
  "1",
  "3",
  "0",
  "4",
  "3",
  "3",
  "7",
  "3",
  "5",
  "2",
  "6",
  "6",
  "4",
  "4",
  "2",
  "1",
  "1",
]
`;

exports[`posts table shows posts ordered by time according to cell that is clicked 1`] = `
Array [
  Object {
    "author": "s_magik",
    "authorHref": "https://reddit.com/u/s_magik",
    "href": "https://reddit.com/r/javascript/comments/k7rxg7/open_source_airtable_built_using_reactjs/",
    "numComments": "21",
    "score": "235",
    "time": "12:08 pm",
    "title": "Open source Airtable: Built using ReactJS + Firestore DB",
  },
  Object {
    "author": "zvikfir",
    "authorHref": "https://reddit.com/u/zvikfir",
    "href": "https://reddit.com/r/javascript/comments/pr5cef/nextjs_nestjs_mobx_and_how_i_structured_my/",
    "numComments": "12",
    "score": "126",
    "time": "12:17 pm",
    "title": "Next.js, NestJS, MobX and how I structured my open-source project",
  },
  Object {
    "author": "Well_Gravity",
    "authorHref": "https://reddit.com/u/Well_Gravity",
    "href": "https://reddit.com/r/javascript/comments/ludsvl/reading_an_excel_file_in_node/",
    "numComments": "7",
    "score": "115",
    "time": "12:31 pm",
    "title": "Reading an Excel file in node",
  },
  Object {
    "author": "GunvantSr",
    "authorHref": "https://reddit.com/u/GunvantSr",
    "href": "https://reddit.com/r/javascript/comments/q55pz1/a_simple_boilerplate_generator_for_your_node/",
    "numComments": "2",
    "score": "254",
    "time": "12:44 pm",
    "title": "A simple boilerplate generator for your node express backend project",
  },
]
`;
